csrc/multidevice/communicator.cpp

```diff
@@ -158,32 +160,54 @@ Communicator::Communicator(
                            master_addr_ == gethostbyname(hostname)->h_name) &&
         local_rank_ == server_local_rank;
   }
-  if (master_port_) {
-    store_opts.port = master_port_;
+  store_opts.port = master_port_ ? master_port_ : comm_master_port_default;
+  store_ = c10::make_intrusive<c10d::TCPStore>(master_addr_, store_opts);
+
+  // creates the world's backend
+  std::vector<RankType> all_ranks(size_);
+  std::iota(all_ranks.begin(), all_ranks.end(), 0);
+  world_ = getBackendForTeam(all_ranks);
+}
+
+c10::intrusive_ptr<c10d::Backend> Communicator::getBackendForTeam(
+    const std::vector<DeviceIdxType>& ranks) {
+  // check if backend associated with the ranks is present in the cache
+  if (teams_.find(ranks) == teams_.end()) { // create the backend and cache it
+    // check that the caller's rank belongs to the requested team
+    auto rank_it = std::find(ranks.begin(), ranks.end(), rank_);
+    NVF_ERROR(
+        rank_it != ranks.end(),
+        "only ranks in the team should participate to its initialization");
+    // retrieve the caller's rank index/position in the team
+    RankType team_rank = std::distance(ranks.begin(), rank_it);
+    // generate a string key which is unique to the team
+    std::string team_key = std::accumulate(
+        std::begin(ranks),
+        std::end(ranks),
+        std::string{},
+        [](const std::string& a, const RankType& b) {
+          return a.empty() ? std::to_string(b) : a + ',' + std::to_string(b);
+        });
+    // create the team and cache it
+    teams_[ranks] = createBackend(
+        backend_,
+        c10::make_intrusive<c10d::PrefixStore>(team_key, store_),
+        team_rank,
+        ranks.size());
   }
-  auto store = c10::make_intrusive<c10d::TCPStore>(master_addr_, store_opts);
-  pg_ = createBackend(backend, store, rank_, size_);
+  return teams_.at(ranks);
 }
 
-void Communicator::sendRecv(
-    RankType receiver_rank,
-    RankType sender_rank,
-    std::vector<at::Tensor>& tensor,
+c10::intrusive_ptr<c10d::Work> Communicator::sendRecv(
+    DeviceIdxType receiver,
+    DeviceIdxType sender,
+    std::vector<at::Tensor>& tensors,
     int tag) {
-  if (sender_rank == receiver_rank) { // send-to-self
-    return;
-  }
-  if (rank() == sender_rank) {
-    // post send and wait for completion
-    NVF_ERROR(
-        pg_->send(tensor, receiver_rank, tag)->wait(),
-        "error during communication");
-  } else if (rank() == receiver_rank) {
-    // post receive and wait for completion
-    NVF_ERROR(
-        pg_->recv(tensor, sender_rank, tag)->wait(),
-        "error during communication");
+  NVF_ERROR(sender != receiver, "cannot send to self");
+  if (deviceId() == sender) {
+    return world_->send(tensors, static_cast<int>(dIdToRank(receiver)), tag);
   }
+  return world_->recv(tensors, static_cast<int>(dIdToRank(sender)), tag);
```

Comment:
Previously, the recv was only posted for the receiving rank, but now it's done always except for the sender process. Is it guaranteed that this function is only called by the sender and receiver processes? If so, why not asserting that `rank() == sender_rank || rank() == receiver_rank`?

Reply:
I added the assertion